query/math.go

```diff
@@ -90,27 +107,39 @@ func (v *MathValuer) Call(name string, args []influxql.Expr) (interface{}, bool)
 		if !ok {
 			return nil, false
 		}
-		
-		var fn func(float64, float64) float64
-		switch name {
-		case "atan2":
-			fn = math.Atan2
-		case "pow":
-			fn = math.Pow
-		}
-		
-		if fn != nil {
-			result := fn(arg0, arg1)
-			if math.IsNaN(result) {
-				return nil, false
-			} else {
-				return result, true
+		return math.Atan2(arg0, arg1), true
+	case "pow":
+		// Take two args. Return a float unless we get an integer raised to a
+		// non-negative integer power.
+		if len(args) != 2 {
+			return nil, false
+		}
+		if arg0, ok := v.getIntegerValue(args[0]); ok {
+			if arg1, ok := v.getIntegerValue(args[1]); ok && arg1 >= 0 {
+				// Do the calc as a float then convert back to int if possible
+				result := math.Pow(float64(arg0), float64(arg1))
```

Comment:
I would prefer if we did the integer power using only integers if possible. It should likely be implemented like this:

```golang
switch arg1 {
case 2:
    return arg0 << 1, true
case 4:
    return arg0 << 2, true
case 8:
    return arg0 << 3, true
case 16:
    return arg0 << 4, true
default:
    val := arg0
    for i := 1; i < arg1; i++ {
        val *= arg0
    }
    return val, true
}
```

This would also include an optimization for bitshift. I'm not sure if the go compiler would get that automatically, but I would guess it would not in this scenario. It might be worth checking the optimizer though.

One advantage of the method used here is that we would be able to detect overflow, but no other integer operation really checks for that so I don't think that's as important as accuracy for the integer type.

As a reminder, absolute value and power need to also be implemented for the unsigned type. For unsigned, if one of the two is unsigned and one is integer, it needs to treat it as unsigned.

Reply:
Hmmm, bitshifting two places to the left would multiply by 4, not raise to the power of 4.  And I suspect that the naive "iterate arg1 times" approach will usually be way slower than converting to float, using the [heavily optimised math library algorithm](https://golang.org/src/math/pow.go)  and converting back to int.  

Maybe we should handle powers of 2 and 3 by using basic multiplication, that should optimise the most likely scenarios I guess.  I'll push a change.

Reply:
Do we try to return a consistent datatype across the calculation?  At the moment we are returning *int* for pow(3, my_int_series) where my_int_series is >= 0 and a float otherwise... depending on the values of the series we could have a mixed return type.   Should we actually only return int if the second argument is a **literal** int >= 0?

Reply:
The number of questions that are coming up because I'm being dumb is becoming problematic so I'll say something different that will hopefully be less dumb.

We should go back to your original implementation and `pow()` should only return float values and then it should be put in the documentation. For anything with `x^2` you can always just do `x * x` if you need to guarantee accuracy as an int. For something like `2^x`, we support bitshifts and I hope that's good enough. We wouldn't have `3^x`, but I think the additional complication caused by this request is creating too big of an issue.

Then we just put it in the documentation. I hope this isn't a bad idea as I cannot predict what people are going to need, but we could add a new power function for influxql if it comes up and we can fix the behavior in ifql.

Sorry for throwing you around so much on this. I didn't realize it would be so complicated to do what I thought.